sys/clang-format-diff.py

```diff
@@ -0,0 +1,152 @@
+                      help='be more verbose, ineffective without -i')
+  parser.add_argument('-style',
+                      help='formatting style to apply (LLVM, Google, Chromium, '
+                      'Mozilla, WebKit)')
+  parser.add_argument('-binary', default='clang-format',
+                      help='location of binary to use for clang-format')
+  args = parser.parse_args()
+
+  # Extract changed lines for each file.
+  filename = None
+  lines_by_file = {}
+  for line in sys.stdin:
+    match = re.search('^\+\+\+\ (.*?/){%s}(\S*)' % args.p, line)
+    if match:
+      filename = match.group(2)
+    if filename == None:
+      continue
+
+    if args.regex is not None:
+      if not re.match('^%s$' % args.regex, filename):
+        continue
+    else:
+      if not re.match('^%s$' % args.iregex, filename, re.IGNORECASE):
+        continue
+
+    match = re.search('^@@.*\+(\d+)(,(\d+))?', line)
+    if match:
+      start_line = int(match.group(1))
+      line_count = 1
+      if match.group(3):
+        line_count = int(match.group(3))
+      if line_count == 0:
+        continue
+      end_line = start_line + line_count - 1
+      lines_by_file.setdefault(filename, []).append(
+          [start_line, end_line])
+
+  # Reformat files containing changes in place.
+  for filename, lines in lines_by_file.items():
+    command = [args.binary, filename]
+    if args.sort_includes:
+      command.append('-sort-includes')
+    if lines:
+        s = [('-lines', str(x[0]) + ':' + str(x[1])) for x in lines]
+        s = reduce(lambda x, y: x + y, s)
+        command.extend(s)
+    if args.style:
+      command.extend(['-style', args.style])
+    p = subprocess.Popen(command,
+                         stdout=subprocess.PIPE,
+                         stderr=None,
+                         stdin=subprocess.PIPE,
+                         universal_newlines=True)
+    stdout, stderr = p.communicate()
+    if p.returncode != 0:
+      sys.exit(p.returncode)
+
+    with open(filename) as f:
+      code = f.readlines()
+    formatted_code = StringIO(stdout).readlines()
+    modified_lines = dict()
+    if lines:
+        for x in lines:
+            for i in range(x[0], x[1] + 1):
+                modified_lines[i] = True
+
+    # handle functions definitions/declarations: do not use space before (
+    for i, l in enumerate(formatted_code):
+        if lines and i + 1 not in modified_lines:
+            continue
+
+        if l.startswith('R_API ') or l.startswith('static '):
+            formatted_code[i] = l.replace(' (', '(')
+
+    diff = difflib.unified_diff(code, formatted_code,
+                                filename, filename,
+                                '(before formatting)', '(after formatting)')
+    diff_string = ''.join(diff)
+    if len(diff_string) > 0:
+      if args.i:
+        f = tempfile.NamedTemporaryFile(delete=False)
+        f.write(diff_string)
+        f.close()
+        os.system('git apply -p0 < %s' % (f.name))
```

Comment:
doesnt work with path containing spaces. better use exec approach and feed stdin with a pipe from python

Reply:
can't i just wrap the file name in `"`? I know it's cleaner to do as you say, but we don't need anything fancy here, just git apply to run.

Reply:
> can't i just wrap the file name in `"`?

That would still be vulnerable to filenames that contain `\` or `"`.

Reply:
but this name is coming from `tempfile.NamedTemporaryFile(delete=False)`, it's not user-supplied, so I don't think it contains those things.

Reply:
> I don't think it contains those things.

Maybe that is a good guess, but https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile does not say anything about the format of the temporary file name.

On the other hand, it says something about the file not being able to be opened a second time on Windows using the provided file name. That may be a problem if you want to support Windows. (Disclaimer: I do not know anything about Windows.)

Reply:
` Whether the name can be used to open the file a second time, while the named temporary file is still open, varies across platforms`. I'm not opening the file a second time *while the temporary file is still open*.